src/khepri_machine.erl

```diff
@@ -1194,6 +1287,89 @@ apply(
                      State
              end,
     Ret = {State1, Reply},
+    bump_applied_command_count(Ret, Meta);
+apply(
+  Meta,
+  #acquire_lock{lock = #khepri_lock{lock_id = LockId,
+                                    group_id = GroupId,
+                                    release_on_exit = ReleaseOnExit},
+                acquirer = Acquirer},
+  #?MODULE{locks = Locks} = State) ->
+    Ret = case maps:get(LockId, Locks, {GroupId, []}) of
+              {GroupId, Acquirers} ->
+                  Acquirer1 = {Acquirer, ReleaseOnExit},
+                  LockEntry = {GroupId, [Acquirer1 | Acquirers]},
+                  Locks1 = Locks#{LockId => LockEntry},
+                  State1 = State#?MODULE{locks = Locks1},
+                  Effects = case ReleaseOnExit of
+                                true ->
+                                    [{monitor, process, Acquirer}];
+                                false ->
+                                    []
+                            end,
+                  {State1, ok, Effects};
+              _OtherGroup ->
+                  {State, {error, unavailable}}
+          end,
+    bump_applied_command_count(Ret, Meta);
+apply(
+  Meta,
+  #release_lock{lock = #khepri_lock{lock_id = LockId,
+                                    group_id = GroupId},
+                releaser = Releaser},
+  #?MODULE{locks = Locks} = State) ->
+    Ret = case Locks of
+              #{LockId := {GroupId, Acquirers}} ->
+                  case lists:keytake(Releaser, 1, Acquirers) of
+                      {value, {Releaser, ReleaseOnExit}, Acquirers1} ->
+                          Locks1 = case Acquirers1 of
+                                       [] ->
+                                           maps:remove(LockId, Locks);
+                                       _OtherAcquirers ->
+                                           LockEntry = {GroupId, Acquirers1},
+                                           Locks#{LockId => LockEntry}
+                                   end,
+                          State1 = State#?MODULE{locks = Locks1},
+                          Effects = case ReleaseOnExit of
+                                        true ->
+                                            [{demonitor, process, Releaser}];
```

Comment:
If we can monitor once each process, the demonitoring will be a bit more complicated because it should be issued after the last release of this process only (if the `release_after_exit` flag was true).